pkg/runtime/runtime.go

```diff
@@ -118,140 +120,41 @@ type TestResult struct {
 	FailedProperty   string
 	Tries            int
 	Node             string
-	Error            error
 	FileName         string
 }
 
 // Start starts the given test suite and executes all tests
-func (r *Runtime) Start(tests []TestCase) <-chan TestResult {
-	in := make(chan TestCase)
-	out := make(chan TestResult)
-
-	go func(tests []TestCase) {
-		defer close(in)
-		for _, t := range tests {
-			in <- t
-		}
-	}(tests)
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-
-	go func(tests chan TestCase) {
-		defer wg.Done()
-
-		for t := range tests {
-			// If no node was set use local mode as default
-			if len(t.Nodes) == 0 {
-				t.Nodes = []string{"local"}
-			}
-
-			for _, n := range t.Nodes {
-				result := TestResult{}
-				for i := 1; i <= t.Command.GetRetries(); i++ {
-
-					e := r.getExecutor(n)
-					result = e.Execute(t)
-					result.Node = n
-					result.Tries = i
-
-					if result.ValidationResult.Success {
-						break
-					}
-
-					executeRetryInterval(t)
-				}
-				out <- result
-			}
-
-		}
-	}(in)
-
-	go func(results chan TestResult) {
-		wg.Wait()
-		close(results)
-	}(out)
-
-	return out
-}
-
-func (r *Runtime) getExecutor(node string) Executor {
-	if len(r.Nodes) == 0 {
-		return NewLocalExecutor()
-	}
-
-	for _, n := range r.Nodes {
-		if n.Name == node {
-			switch n.Type {
-			case "ssh":
-				return NewSSHExecutor(n.Addr, n.User, WithIdentityFile(n.IdentityFile), WithPassword(n.Pass))
-			case "local":
-				return NewLocalExecutor()
-			case "docker":
-				log.Println("Use docker executor")
-				return DockerExecutor{
-					Image:        n.Image,
-					Privileged:   n.Privileged,
-					ExecUser:     n.DockerExecUser,
-					RegistryPass: n.Pass,
-					RegistryUser: n.User,
-				}
-			case "":
-				return NewLocalExecutor()
-			default:
-				log.Fatal(fmt.Sprintf("Node type %s not found for node %s", n.Type, n.Name))
-			}
+func (r *Runtime) Start(tests []TestCase) output.Result {
```

Comment:
Your implementation is heading in the right direction. 
At first I pointed out that the output should not now anything about the `runtime`.
Now maybe it would be nicer if the runtime doesn't know anything about the output, atm it is highly depends on the `output` package because 
1. it is called directly
2. return an `output.Result` as the aggregated result, which is a little bit weird because the runtime package always uses it's own structs - and suddenly it uses the `output.Result` because of circular dependencies (which a normal dev would never see)

My proposal in this case would be to integrate a light weight "event"-system which is just a struct which holds some functions which are called at a specific time, i.e. a test finished.
In Kubernetes this would be called a `SharedInformer`, I would like to stick to `EventHandler` because it is more understandable.

```go
type EventHandler struct {
    TestFinsihed func(TestCase)
    RuntimeFinished func(Result)
}

// Start starts the given test suite and executes all tests
func (r *Runtime) Start(tests []TestCase) Result {
	testCh := r.Runner.Execute(tests)
       result := Result{} // Result is defined like the output.Result but located inside the runtime
	start := time.Now()
	for tr := range testCh {
		tr := convertTestResult(tr)
	        // Instead of printing the result the TestFinished function gets called
                //r.Output.PrintResult(tr)
               r.EventHandler.TestFinished(tr)

		if !tr.Success {
			result.Failed++
		}

		result.TestResults = append(result.TestResults, tr)
	}
	result.Duration = time.Since(start)
        
        r.EventHandler.RuntimeFinished()
	return result
}
```

It would be possible to add more listeners, even adding a slice of listeners allowing multiple functions to hook into the execution.
Now if an output should be configured it could be by adding the event handlers, i.e. inside the `output` package
```go
func NewEventHandler() runtime.EventHandler {
   testFunc := function (testResult runtime.TestResult) {
       // printResult...
   }

   finishFunc := function (result runtime.Result) {
       // printSummary...
   }
}
```

The `EventHandler` could be added as a property to the `runtime.Runtime` struct. 
What are your thoughts on this? 
This should make testing easier too.

Reply:
I like the idea of an EventListener. I don't think we quite have a use for `r.EventHandler.RuntimeFinished()` this yet, but i see where you're going. You did lose me  at the very end and I think it's mainly I'm losing where structs should be defined. So:

```golang
type EventHandler struct {
    TestFinsihed func(TestCase)
    RuntimeFinished func(Result)
}
```
^This is defined in `runtime`

```golang
// convert runtime.TestResult to output.TestResult
func convertTestResult(tr TestResult) output.TestResult {
	testResult := output.TestResult{
		FileName:       tr.TestCase.FileName,
		Title:          tr.TestCase.Title,
		Node:           tr.Node,
		Tries:          tr.Tries,
		Success:        tr.ValidationResult.Success,
		FailedProperty: tr.FailedProperty,
		Diff:           tr.ValidationResult.Diff,
		Error:          tr.TestCase.Result.Error,
	}

	return testResult
}

```
^this will be moved to output in some capacity. Ideally there will be some conversion that takes place in `r.EventHandler.TestFinished(tr)`

```golang
func NewEventHandler() runtime.EventHandler {
   testFunc := function (testResult runtime.TestResult) {
       // printResult...
   }

   finishFunc := function (result runtime.Result) {
       // printSummary...
   }
}
```
^this code would belong in `output`? If this code is in output. We would need something on these lines in app
```golang
r := runtime.NewRuntime(out.NewEventHander, s.Nodes...)
result := r.Start(tests)
```
and runtime is 
```golang
type Runtime struct {
	Runner *Runner
	event     EventHandler
}
```

I was. going to suggest that eventhandler should get it's own package rather than being generated over in `output`.  However if we ever start supporting some other `out` (; each `output` should be responsible for generating its own `eventhandler`. <- you may have already known that and that's why you suggested doing it this way. I really like that solution.

Let me know if I followed you correctly.

Best

Reply:
Yes you followed correctly! 
Indeed the `output` should be responsible for creating it's own `EventHandler`. 
If we want to support more outputs we could add a global variable which maintains a registry of all available output formats inside a hashmap, i.e.

```
// runtime.go
// {"cli": CLIEventHandler, "json": JSONEventHandler}
var OutputHandlers map[string]*EventHandler

func RegisterOutput(name string, eventHandler *EventHandler) {
     OutputHandlers[name] = eventHandler
}
``` 

The `cli.go` would register the `EventHandler` in an `init` func.

```
// cli.go
func init() {
    runtime.RegisterOutput(eventHandler)
}
```

Reply:
Cool, I hope to get this finished up by Tuesday evening.

Thanks again!

Reply:
I plan on getting to this, finding the time lately has been difficult.

Reply:
@SimonBaeumer the EventHandler made all this feel super nice. I didn't really focus on making it more configurable yet. I think that would be a task suited for when we have second `output` type. I figure it would something a long the lines of requiring all `outs` to implement an interface, but that's for a later PR.